usr/src/uts/common/fs/zfs/sys/spa.h

```diff
@@ -326,7 +327,8 @@ typedef enum bp_embedded_type {
 typedef struct blkptr {
 	dva_t		blk_dva[SPA_DVAS_PER_BP]; /* Data Virtual Addresses */
 	uint64_t	blk_prop;	/* size, compression, type, etc	    */
-	uint64_t	blk_pad[2];	/* Extra space for the future	    */
+	uint32_t	blk_iv[3];	/* IV for block level encryption    */
+	uint32_t	blk_pad;	/* Extra space for the future	    */
```

Comment:
> the bookmark actually won't work. A snapshot will have a different objset 

Yeah, I guess you'd want something like a "clone family ID". (Or somewhat equivalently, the "key ID".)

> The attacker would first need the key to already be loaded into the pool, otherwise the dataset can't write to do dedup in the first place anyway. They would also need write access to the dataset, at which point they can just read the data anyway.

I'm imagining a scenario where I can send you data to write to the pool (e.g. I upload my photos to your website) and then later I steal / access your hard drives.

If that isn't a concern, then wouldn't it be OK to always use the same IV?  It would certainly simplify things.